File: crates/astar-sys/examples/parrot.c
/*
 * parrot.c — a minimal C consumer of the astar-sys poll+snapshot C-ABI.
 *
 * Demonstrates the full lifecycle with NO callbacks: new -> connect (the
 * generic, vendor-neutral path) -> poll snapshot/next_event in a loop ->
 * disconnect -> free.
 *
 * Secret-free by construction: the guest secret is passed as a connect()
 * IN-param ("allstar" here) and is never read back out of any struct.
 *
 * Build + link: see examples/build.sh and examples/README.md.
 *
 * Running it for real dials the AllStar parrot 55553 (audio loopback). The CI
 * job only compiles + links this file; it does not run it (no network in CI).
 * Define IAX_PARROT_DRYRUN to build a self-contained offline smoke that just
 * exercises new -> snapshot -> set_ptt(NOT_CONNECTED) -> free.
 */

#include "astar.h"

#include <stdio.h>
#include <string.h>

#if defined(_WIN32)
#  include <windows.h>
static void sleep_ms(int ms) { Sleep(ms); }
#else
#  include <unistd.h>
static void sleep_ms(int ms) { usleep((useconds_t)ms * 1000); }
#endif

static const char *status_name(IaxStatus s) {
    switch (s) {
        case IaxStatus_Idle:     return "idle";
        case IaxStatus_Dialing:  return "dialing";
        case IaxStatus_Answered: return "answered";
        case IaxStatus_Hangup:   return "hangup";
        default:                 return "?";
    }
}

static void drain_events(IaxStation *st) {
    for (;;) {
        IaxEvent ev;
        int rc = iax_station_next_event(st, &ev);
        if (rc != IAX_OK || ev.kind == IaxEventKind_None) {
            return;
        }
        switch (ev.kind) {
            case IaxEventKind_Answered:
                printf("  event: answered\n");
                break;
            case IaxEventKind_RemotePtt:
                printf("  event: remote_ptt=%d\n", ev.remote_ptt ? 1 : 0);
                break;
            case IaxEventKind_Hangup:
                printf("  event: hangup\n");
                break;
            default:
                break;
        }
    }
}

int main(void) {
    /* All NULL: system default devices, no portal (manual/parrot path), and
     * the default guest secret. */
    IaxConfig cfg;
    memset(&cfg, 0, sizeof cfg);

    IaxStation *st = iax_station_new(&cfg);
    if (!st) {
        fprintf(stderr, "iax_station_new failed\n");
        return 1;
    }

#ifdef IAX_PARROT_DRYRUN
    /* Offline smoke: no network. Prove the poll surface and the secret-free
     * not-connected guard, then free. */
    IaxState s;
    int rc = iax_station_snapshot(st, &s);
    if (rc != IAX_OK || s.status != IaxStatus_Idle) {
        fprintf(stderr, "expected idle snapshot, rc=%d status=%d\n", rc, s.status);
        iax_station_free(st);
        return 1;
    }
    rc = iax_station_set_ptt(st, true);
    if (rc != IAX_ERR_NOT_CONNECTED) {
        fprintf(stderr, "expected NOT_CONNECTED, got %d (%s)\n", rc, iax_error_text(rc));
        iax_station_free(st);
        return 1;
    }
    /* send_dtmf smoke: an off-pad character is rejected as an invalid digit
     * (before any call check), and a valid key with no call is not-connected. */
    rc = iax_station_send_dtmf(st, 'A');
    if (rc != IAX_ERR_INVALID_DIGIT) {
        fprintf(stderr, "expected INVALID_DIGIT, got %d (%s)\n", rc, iax_error_text(rc));
        iax_station_free(st);
        return 1;
    }
    rc = iax_station_send_dtmf(st, '5');
    if (rc != IAX_ERR_NOT_CONNECTED) {
        fprintf(stderr, "expected NOT_CONNECTED for dtmf, got %d (%s)\n", rc, iax_error_text(rc));
        iax_station_free(st);
        return 1;
    }
    printf("dry run ok: idle snapshot + not-connected guard + dtmf guards\n");
    iax_station_free(st);
    return 0;
#else
    /* Generic IAX2 connect (vendor-neutral): dial the parrot. secret is an
     * in-param, consumed and never echoed. */
    int rc = iax_station_connect(st, "55553", "55553", "allstar", "astar");
    if (rc != IAX_OK) {
        fprintf(stderr, "connect failed: %s (%d)\n", iax_error_text(rc), rc);
        iax_station_free(st);
        return 1;
    }
    printf("connected; polling ~8s...\n");

    for (int i = 0; i < 80; i++) {
        IaxState s;
        if (iax_station_snapshot(st, &s) == IAX_OK) {
            printf("status=%-8s tx=%.1fdB rx=%.1fdB ptt=%d rtt=%dms\n",
                   status_name(s.status), s.tx_db, s.rx_db, s.ptt ? 1 : 0, s.rtt_ms);
        }
        drain_events(st);
        sleep_ms(100);
    }

    iax_station_disconnect(st);
    iax_station_free(st);
    printf("done\n");
    return 0;
#endif
}

File: bindings/python/examples/parrot.py
#!/usr/bin/env python3
# astar — Copyright (c) 2026 Rob Ludwick.
# SPDX-License-Identifier: AGPL-3.0-only
# Licensed under the GNU Affero General Public License v3.0 only. See LICENSE.
"""parrot.py — Python consumer of the astar-sys poll+snapshot C-ABI.

Mirrors crates/astar-sys/examples/parrot.c: new -> connect (the generic,
vendor-neutral path) -> poll snapshot/next_event in a loop -> disconnect ->
free. No callbacks.

Secret-free by construction: the guest secret is a connect() IN-param
("allstar" here) and is never read back out of any snapshot/event/object.

Usage::

    python3 examples/parrot.py            # dials the AllStar parrot 55553
    python3 examples/parrot.py --dry-run  # offline smoke, no network

The --dry-run path needs no network: new -> idle snapshot ->
set_ptt(NOT_CONNECTED) -> next_event(None) -> free, with a secret-free guard.
"""

import os
import sys
import time

# Allow `python3 examples/parrot.py` from anywhere: import the sibling module.
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))

from astarstation import (  # noqa: E402
    IAX_ERR_NOT_CONNECTED,
    Station,
    StationError,
    Status,
)


def dry_run() -> int:
    """Offline smoke: prove the poll surface + not-connected guard. No network."""
    secret = "allstar"  # in-param only; must never surface anywhere readable.
    with Station() as st:
        snap = st.snapshot()
        if snap.status != Status.IDLE:
            print(f"expected idle snapshot, got {snap.status}", file=sys.stderr)
            return 1
        print(f"snapshot: status={snap.status.name} tx={snap.tx_db}dB rx={snap.rx_db}dB")

        try:
            st.set_ptt(True)
        except StationError as e:
            if e.code != IAX_ERR_NOT_CONNECTED:
                print(f"expected NOT_CONNECTED, got {e.code} ({e.text})", file=sys.stderr)
                return 1
            print(f"set_ptt(True) while idle -> {e}")
        else:
            print("set_ptt(True) should have raised while idle", file=sys.stderr)
            return 1

        if st.next_event() is not None:
            print("expected no event while idle", file=sys.stderr)
            return 1
        print("next_event() -> None")

        # Secret-free guard: the in-param secret must not appear anywhere the
        # app can read.
        for surface in (repr(st), str(snap)):
            if secret in surface or "secret" in surface.lower():
                print(f"secret leaked in: {surface!r}", file=sys.stderr)
                return 1
        print("secret-free guard: ok")

    print("dry run ok")
    return 0


def live(node: str = "55553") -> int:
    """Dial the parrot and poll for ~8s. Requires network + audio hardware."""
    with Station() as st:
        # Generic IAX2 connect (vendor-neutral). secret is an in-param.
        try:
            st.connect(node, node, secret="allstar", name="astar")
        except StationError as e:
            print(f"connect failed: {e}", file=sys.stderr)
            return 1
        print("connected; polling ~8s...")

        for _ in range(80):
            snap = st.snapshot()
            print(
                f"status={snap.status.name:<8} tx={snap.tx_db:.1f}dB "
                f"rx={snap.rx_db:.1f}dB ptt={int(snap.ptt)} "
                f"rtt={snap.rtt_ms if snap.rtt_ms is not None else -1}ms"
            )
            while (ev := st.next_event()) is not None:
                if ev.kind.name == "REMOTE_PTT":
                    print(f"  event: remote_ptt={int(ev.remote_ptt)}")
                else:
                    print(f"  event: {ev.kind.name.lower()}")
            time.sleep(0.1)

        st.disconnect()
    print("done")
    return 0


def main(argv: list[str]) -> int:
    if "--dry-run" in argv or os.environ.get("IAX_PARROT_DRYRUN") == "1":
        return dry_run()
    return live()


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))
